pageserver/src/tenant_mgr.rs

```diff
@@ -59,153 +54,79 @@ mod tenants_state {
+                    );
+                    if let Err(e) = std::fs::remove_dir_all(&tenant_dir_path) {
+                        error!(
+                            "Failed to remove temporary directory '{}': {:?}",
+                            tenant_dir_path.display(),
+                            e
+                        );
                     }
-                    TenantAttachData::Ready(_) => {
-                        match Tenant::load_tenant_config(conf, tenant_id) {
-                            Ok(tenant_conf) => {
-                                tenant.update_tenant_config(tenant_conf);
-                                tenant.activate(false);
-                            }
-                            Err(e) => {
-                                error!("Failed to read config for tenant {tenant_id}, disabling tenant: {e:?}");
-                                tenant.set_state(TenantState::Broken);
-                            }
-                        };
+                } else {
+                    number_of_tenants += 1;
+                    if let Err(e) =
+                        load_local_tenant(conf, &tenant_dir_path, remote_storage.clone())
+                    {
+                        error!(
+                            "Failed to collect tenant files from dir '{}' for entry {:?}, reason: {:#}",
+                            tenants_dir.display(),
+                            dir_entry,
+                            e
+                        );
                     }
                 }
-                v.insert(Arc::clone(&tenant));
-                tenant
             }
-        };
-        drop(tenants_accessor);
-        match local_timelines {
-            TenantAttachData::Broken(e) => warn!("{}", e),
-            TenantAttachData::Ready(ref timelines) => {
-                info!("Attaching {} timelines for {tenant_id}", timelines.len());
-                debug!("Timelines to attach: {local_timelines:?}");
-                let has_timelines = !timelines.is_empty();
-                let timelines_to_attach = timelines
-                    .iter()
-                    .map(|(&k, v)| (k, v.metadata().to_owned()))
-                    .collect();
-                match tenant.init_attach_timelines(timelines_to_attach) {
-                    Ok(()) => {
-                        info!("successfully loaded local timelines for tenant {tenant_id}");
-                        tenant.activate(has_timelines);
-                    }
-                    Err(e) => {
-                        error!("Failed to attach tenant timelines: {e:?}");
-                        tenant.set_state(TenantState::Broken);
-                    }
-                }
+            Err(e) => {
+                // On error, print it, but continue with the other tenants. If we error out
+                // here, the pageserver startup fails altogether, causing outage for *all*
+                // tenants. That seems worse.
```

Comment:
Yeah, that's a problem.

In general, if we fail to initialize one tenant, we try to proceed with others anyway, like the new comment says. We used to error out and die completely, but we actually had a case in production where there was a problem with one tenant and that caused an outage for everyone. We added the Broken tenant mechanism because of that. We don't create a Broken tenant in this readdir() failure case, though. Getting an error from readdir() would be quite extraordinary, so I tend to agree that it would actually be better to just die and refuse to start up if that happens.

This isn't new with this PR, the added comment just documents the old behaviour, so I think this should be changed as a separate PR.

Reply:
Tracked by https://github.com/neondatabase/neon/issues/2933